plugins/logitech-rallysystem/fu-logitech-rallysystem-plugin.c

```diff
@@ -21,6 +23,51 @@ fu_logitech_rallysystem_plugin_init(FuLogitechRallysystemPlugin *self)
 {
 }
 
+static void
+fu_logitech_rallysystem_plugin_finalize(GObject *object)
+{
+	FuLogitechRallysystemPlugin *self = FU_LOGITECH_RALLYSYSTEM_PLUGIN(object);
+	g_free(self->system_version);
+	G_OBJECT_CLASS(fu_logitech_rallysystem_plugin_parent_class)->finalize(object);
+}
+
+static void
+fu_logitech_rallysystem_plugin_device_added(FuPlugin *plugin, FuDevice *device)
+{
+	FuLogitechRallysystemPlugin *self = FU_LOGITECH_RALLYSYSTEM_PLUGIN(plugin);
+	GPtrArray *devices;
+
+	/*
+	 * composite device is composed of multiple sub-devices: audio, video, tablehub, speakers.
+	 * Each sub-device has their own unique firmware version.
+	 * Only audio sub-device has topology information of all sub-devices.
+	 * Only audio sub-device has information of overall/unique system version.
+	 * Only tablehub device is responsible to push firmware images to all sub-devices,
+	 * all sub-devices update firmware independently, and provide status to tablehub.
+	 * Since only tablehub can accept firmware image, its guid is used in .metainfo.xml file.
+	 * Ask here is for tablehub, to provide system version, so that application can use single
+	 * guid to query system version and check .metainfo.xml to determine if upgraded is needed
+	 * or not. Following logic preserve the system version information from audio sub-device and
+	 * overwrite local version information of tablehub with this system version
+	 */
+	if (g_strcmp0(fu_device_get_plugin(device), "logitech_rallysystem") == 0) {
+		if (FU_IS_LOGITECH_RALLYSYSTEM_AUDIO_DEVICE(device)) {
+			self->system_parent_plugin = plugin;
+			self->system_version = strdup(fu_device_get_version(device));
+		}
+		devices = fu_plugin_get_devices(plugin);
+		for (guint i = 0; i < devices->len; i++) {
+			FuDevice *device_tmp = g_ptr_array_index(devices, i);
+			if ((FU_IS_LOGITECH_RALLYSYSTEM_TABLEHUB_DEVICE(device_tmp)) &&
+			    (self->system_parent_plugin == plugin)) {
+				fu_device_set_version(device_tmp, self->system_version);
+				g_debug("overwriting tablehub version to %s", self->system_version);
+				break;
+			}
```

Comment:
can't you do it without even _storing_ to the  plugin?   Just keep it all in scope?

Reply:
How do we pass information between sub-devices? We also need to ensure both sub-devices exist. Do it as part of setup, before write_firmware(). Open for suggestion.

Reply:
I think what @superm1 is saying is that rather than storing it in the plugin, you could do what you're doing here but instead *find* the audio device and tablehub and use `fu_device_set_version(device_tablehub, fu_device_get_version(device_audio))`

Reply:
Yup, that's exactly what I meant.  I just didn't explain it well enough :)